Provide shared context for common coin selection types.

jonathanknowles opened the pull request:
# Issue Number

ADP-1448

# Summary

This PR:

- [x] Parameterizes the type of **_address_** within coin selection modules.
- [x] Introduces a **_shared context_** for common types used within coin selection modules.

## Details

We currently represent address and UTxO identifiers with **_type parameters_** on coin selection-related functions and record types. In the future, we also intend to make coin selection agnostic to other types, such as the type of asset identifiers.

Many of these type parameters are repeated across multiple record types and functions, which leads to quite a bit of redundancy and arguably makes the code harder to read and edit.

To alleviate this problem, this PR introduces a new `SelectionContext` class, which will eventually have at least three associated types:

```hs
class
    ( Ord (Address c)
    , Ord (Asset c)
    , Ord (UTxO c)
    , ...
    ) =>
    SelectionContext c
  where
    -- | A target address to which payments can be made.
    type Address c
    -- | A unique identifier for an asset.
    type Asset c
    -- | A unique identifier for an individual UTxO.
    type UTxO c
```

This shared context allows us to make simplifications similar to the following:
```patch
- type PerformSelection m f address asset utxo  =
+ type PerformSelection m f ctx                 =
-     SelectionConstraints  address asset utxo ->
+     SelectionConstraints  ctx                ->
-     SelectionParamsOf   f address asset utxo ->
+     SelectionParamsOf   f ctx                ->
      m (Either
-             (SelectionBalanceError address asset utxo)
+             (SelectionBalanceError ctx               )
-             (SelectionResultOf   f address asset utxo)
+             (SelectionResultOf   f ctx               )
      )
```

This PR also provides `WalletSelectionContext`, which is an implementation of `SelectionContext` with wallet-friendly types.

## Future Work

This PR leaves the test suite largely untouched, and continues to use generators and shrinkers for wallet-specific types. Future PRs will convert the test suite to use abstract types.

jonathanknowles committed:
Parameterize the type of address in module `Balance`.

jonathanknowles committed:
Parameterize the type of address in module `CoinSelection.Internal`.

jonathanknowles committed:
Use type parameter `f` for foldable containers in `Balance`.

jonathanknowles committed:
Introduce class `SelectionContext` to group common coin selection types.

For the moment, `SelectionContext` will only have two associated types:

    - Address
      represents target addresses to which payments can be made.

    - UTxO
      represents unique identifiers for individual UTxOs.

A future commit will add an associated type to represent asset identifiers.

jonathanknowles committed:
Update nix.

jonathanknowles committed:
Introduce class `Dummy` to facilitate creation of dummy values.

Within the coin selection algorithm we currently rely on dummy address
values in a couple of places.

Since internal coin selection modules no longer have access to the
concrete type of address (which is now an abstract type parameter), we
can no longer construct these dummy values ourselves.

To solve this problem we provide the `Dummy` class, along with an
instance for `Address`, so that internal coin selection modules can
create these dummy values safely.

jonathanknowles committed:
Add documentation comments.

jonathanknowles commented on lib/core/src/Cardano/Wallet/CoinSelection/Internal.hs:
```diff
@@ -227,30 +227,48 @@ data SelectionParams u = SelectionParams
+    , minimumSelectionAmount :: Coin
+        -- ^ A lower bound on the sum of coins to be selected as collateral.
+    }
+    deriving Generic
```
I've **_forked_** this type to avoid making any changes to the `Internal.Collateral` module, which only needs a single type parameter `u`: the type of unique UTxO identifiers. We **_could_** make the `Collateral` module use `SelectionContext`, but it would add more complication without much benefit.

jonathanknowles commented on lib/core/src/Cardano/Wallet/CoinSelection/Internal/Balance.hs:
```diff
@@ -816,32 +814,38 @@ performSelection = performSelectionEmpty performSelectionNonEmpty
+            transform (`Coin.add` minCoin) (const id) extraCoinSource
+        , outputsToCover =
+            transform (const (dummyOutput :| [])) (const . id) outputsToCover
+        }
```
I changed this section of code to use ordinary record update syntax rather than using `over` (from `generic-lens`).

For some reason that I'm currently unable to ascertain, `over` doesn't seem to play well with type families. I'll investigate changing this back in a future PR.

jonathanknowles commented on lib/core/src/Cardano/Wallet/CoinSelection/Internal/Balance.hs:
```diff
@@ -816,32 +814,38 @@ performSelection = performSelectionEmpty performSelectionNonEmpty
-    dummyOutput :: (Address, TokenBundle)
-    dummyOutput = (Address "", TokenBundle.fromCoin minCoin)
+    dummyOutput :: (Address ctx, TokenBundle)
+    dummyOutput = (dummy, TokenBundle.fromCoin minCoin)
```
Since we no longer have access to the concrete type of address, we can't create address values directly. So we rely on the `Dummy` constraint here.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/CoinSelection/Internal/BalanceSpec.hs:
```diff
@@ -1203,20 +1215,28 @@ prop_performSelectionEmpty mockConstraints (Small params) =
-            [ params === (paramsTransformed & over #outputsToCover F.toList)
-            , resultTransformed === (result & over #outputsCovered F.toList)
+            [ params === paramsTransformed'
+            , resultTransformed === result'
```
Similar to above, this section of code is changed to use ordinary record update syntax instead of `over` (from `generic-lens`).

Anviking commented on lib/core/src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -3,6 +3,10 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
```
This shouldn't be needed?

jonathanknowles replied:
> This shouldn't be needed?

Well spotted!

It was needed due to:
1. The orphan `Buildable` instance for the `InputId` type synonym (which is just a pair of `TxIn` and `Address`).
2. The orphan `Dummy` instance for `Address`.

However, I've added a couple of commits that allow us to remove the `-fno-warn-orphans` statement:

- 1b7bf0c1b2185f91bbdbe432e115d7a8e4dcafb6 (Introduce type `WalletUTxO` to replace `InputId` type synonym).
- 13a090a6e4f7d75fc8d90683d00b6f2f403c5004 (Introduce type `WalletAddress`).

This required me to make a **_slight_** tweak to the `balanceTx` golden test data:

- de158de4c85fee13cc5b4bafd33b623b0b002b3e

Anviking replied:
👌What do you think about ec814b840a3a47aed2ec59385e3f49aea1095970 ? (I pushed to another branch)

By merging the `Dummy` class into `WalletSelectionContext` we can drop the `WalletAddress` wrapper:

```patch
 instance SC.SelectionContext WalletSelectionContext where
-    type Address WalletSelectionContext = WalletAddress
+    type Address WalletSelectionContext = Address
     type UTxO WalletSelectionContext = WalletUTxO

---------------------------------------------------------------------------------
--- Mapping between external (wallet) and internal addresses
---------------------------------------------------------------------------------
-
-newtype WalletAddress = WalletAddress
-    { unWalletAddress :: Address }
-    deriving (Buildable, Eq, Generic, Ord, Show)
-
-instance SC.Dummy WalletAddress where
-    dummy = WalletAddress $ Address ""
+    dummyAddress _ = Address ""
```
to avoid some boilerplate.

We could technically do something similar with `WalletUTxO` and `Buildable`, but that would be more questionable, I think.

jonathanknowles replied:
@Anviking Thanks! I've cherry-picked your commit. (Hope that's okay.)

I think your change definitely simplifies things a bit: we no longer need to wrap and unwrap `Address` values, and we no longer need to use `shrinkMapBy` in order to lift the shrinker for type `Address`.

I was initially surprised that we couldn't remove the `Proxy c` and just write the following definition, and insert type applications in the right places:
```hs
dummyAddress :: Address c
```

But this doesn't work, as `Address` is a non-injective type family. So I've left it as is.

Thanks for writing this commit!

jonathanknowles replied:
It turns out it is possible to avoid the use of `Proxy` here.

See c68133ba8cbd5b906bd40f15879e2463739131cf.

jonathanknowles committed:
Introduce type `WalletUTxO` to replace `InputId` type synonym.

This allows us to remove orphan and overlapping instances for `InputId`.

jonathanknowles committed:
Introduce type `WalletAddress`.

This allows us to define a non-orphan `Dummy` instance for addresses.

jonathanknowles committed:
Tweak `balanceTx` golden test data to account for usage of `WalletUTxO`.

This fix adjusts the golden test data to account for the change in type
from `InputId` (which was just a synonym for a pair of `TxIn` and
`Address`) to `WalletUTxO` (which is a record type).

Anviking reviewed (approved):
👌I think this makes sense

Question: do you expect to add things like
```haskell
    type TxOut c
    output  :: Proxy c -> UTxO -> TxOut
    outputAddress :: Proxy c -> TxOut c -> Address c
    outputValue :: Proxy c -> TxOut c -> TokenBundle
```
to `SelectionContext` in the future? It's doesn't make coin-selection any more abstract, but _perhaps_ it could make it more readable, and lessen the need for wrapping types and functions in `Cardano.Wallet.CoinSelection`?

Anviking committed:
Remove need for `WalletAddress` wrapper.

by merging the `Dummy` class into `SelectionContext`.

jonathanknowles committed:
Remove redundant brackets around `&&&` combinator operands.

Change `(a) &&& b` to `a &&& b`.

(and other minor tweaks to layout)

jonathanknowles committed:
Add comment for function `dummyAddress`.
